src/server/spatial_map.cpp

```diff
@@ -0,0 +1,172 @@
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "spatial_map.h"
+#include <algorithm>
+
+namespace server
+{
+
+// all inserted entires go into the uncached vector
+void SpatialMap::insert(u16 id, const v3f &pos, bool postIteration)
+{
+	if (m_iterators_stopping_insertion_and_deletion) {
+		m_pending_inserts.insert(SpatialKey(pos, id));
+		return;
+	}
+
+	SpatialKey key(pos);
+	if (postIteration)
+		key = SpatialKey(pos.X, pos.Y, pos.Z, true);
+	m_cached.insert({key, id});
+}
+
+// Invalidates upon position update
+void SpatialMap::updatePosition(u16 id, const v3f &oldPos, const v3f &newPos)
+{
+	// Try to leave early if already in the same bucket:
+	auto range = m_cached.equal_range(SpatialKey(newPos));
+	for (auto it = range.first; it != range.second; ++it) {
+		if (it->second == id) {
+			return; // all good, let's get out of here
+		}
+	}
+
+	remove(id, oldPos); // remove from old cache position
+	insert(id, newPos); // reinsert
+}
+
+void SpatialMap::remove(u16 id, const v3f &pos, bool postIteration)
+{
+	if (m_iterators_stopping_insertion_and_deletion) {
+		m_pending_deletes.insert(SpatialKey(pos, id));
+		return;
+	}
+
+	SpatialKey key(pos);
+	if (postIteration)
+		key = SpatialKey(pos.X, pos.Y, pos.Z, true);
+	if(m_cached.find(key) != m_cached.end()) {
+		auto range = m_cached.equal_range(key);
+		for (auto it = range.first; it != range.second; ++it) {
+			if (it->second == id) {
+				m_cached.erase(it);
+				return; // Erase and leave early
+			}
+		}
+	}
+
+	remove(id); // should never be hit
+}
+
+void SpatialMap::remove(u16 id)
+{
+	if(m_iterators_stopping_insertion_and_deletion) {
+		m_pending_deletes.insert(SpatialKey(v3f(), id));
+		return;
+	}
+
+	for (auto it = m_cached.begin(); it != m_cached.end(); ++it) {
+		if (it->second == id) {
+			m_cached.erase(it);
+			return; // Erase and leave early
+		}
+	}
+}
+
+void SpatialMap::removeAll()
+{
+	if(m_iterators_stopping_insertion_and_deletion) {
+		m_remove_all = true;
+	} else {
+		m_cached.clear();
+	}
+}
+
+void SpatialMap::getRelevantObjectIds(const aabb3f &box, const std::function<void(u16 id)> &callback)
+{
+	if (m_cached.empty()) return;
+
+	// When searching, we must round to maximum extent of relevant mapblock indexes.
+	// Since we're using floats, always assume +-1
+	auto low = [](f32 val) -> s16 {
+		s16 _val = static_cast<s16>(val / BS);
+		return (_val >> 4) - 1;
```

Comment:
I wonder if the caller might ever know the exact number of map blocks they intend to check.

Reply:
If you mean they provide a hint/max number to check, that seems like a micro-optimization that won't yield any useable performance gains vs efforts in optimizing elsewhere (once this isn't the bottleneck, that would be collision code next).

Reply:
No I just meant they might have an area defined in terms of map block size, not game coordinates.

Reply:
Ah, hurm. That would require lua API exposure. I mean, they might want to... But then again if they are that smart, they could just provide the correct AABB for it.. I'll leave this comment open until the PR is merged, maybe someone will want to make a feature request actually for such a feature.